packages/@aws-cdk/aws-stepfunctions/lib/state-machine-fragment.ts

```diff
@@ -0,0 +1,81 @@
+import cdk = require('@aws-cdk/cdk');
+import { Chain } from './chain';
+import { Parallel, ParallelProps } from './states/parallel';
+import { State } from './states/state';
+import { IChainable, INextable } from "./types";
+
+/**
+ * Base class for reusable state machine fragments
+ */
+export abstract class StateMachineFragment extends cdk.Construct implements IChainable {
+    /**
+     * The start state of this state machine fragment
+     */
+    public abstract readonly startState: State;
+
+    /**
+     * The states to chain onto if this fragment is used
+     */
+    public abstract readonly endStates: INextable[];
+
+    constructor(parent: cdk.Construct, id: string) {
+        super(parent, id);
+    }
+
+    /**
+     * Prefix the IDs of all states in this state machine fragment
+     *
+     * Use this to avoid multiple copies of the state machine all having the
+     * same state IDs.
+     *
+     * @param prefix The prefix to add. Will use construct ID by default.
+     */
+    public prefixStates(prefix?: string): StateMachineFragment {
+        State.prefixStates(this, prefix || `${this.id}: `);
+        return this;
+    }
+
+    /**
+     * Wrap all states in this state machine fragment up into a single state.
+     *
+     * This can be used to add retry or error handling onto this state
+     * machine fragment.
+     *
+     * Be aware that this changes the result of the inner state machine
+     * to be an array with the result of the state machine in it. Adjust
+     * your paths accordingly. For example, change 'outputPath' to
+     * '$[0]'.
+     */
+    public asSingleState(options: SingleStateOptions = {}): Parallel {
```

Comment:
Maybe rename to `toSingleState` (we try to educate people that the `asXxx` semantics are for interface implementors, and usually shouldn't be called directly by users)